### PR TITLE
Ember Data fixes

### PR DESCRIPTION
**Pull request by chriskrycho**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
    - `DS.Model.serialize`: https://github.com/emberjs/data/blob/b2a157b6500ea1cbf256ae2295096f9900773cd3/addon/-private/system/model/model.js#L428:L444
    - `DS.Model.toJSON`: https://github.com/emberjs/data/blob/b2a157b6500ea1cbf256ae2295096f9900773cd3/addon/-private/system/model/model.js#L446:L466
    - `DS.Model.deleteRecord`: https://github.com/emberjs/data/blob/b2a157b6500ea1cbf256ae2295096f9900773cd3/addon/-private/system/model/model.js#L573:L575
    - `DS.Model.destroyRecord`: https://github.com/emberjs/data/blob/b2a157b6500ea1cbf256ae2295096f9900773cd3/addon/-private/system/model/model.js#L577:L624
    - `DS.Model.unloadRecord`: https://github.com/emberjs/data/blob/b2a157b6500ea1cbf256ae2295096f9900773cd3/addon/-private/system/model/model.js#L626:L636
    - `DS.Model.rollbackAttributes`: https://github.com/emberjs/data/blob/b2a157b6500ea1cbf256ae2295096f9900773cd3/addon/-private/system/model/model.js#L709:L722
    - `DS.Model.save`: https://github.com/emberjs/data/blob/b2a157b6500ea1cbf256ae2295096f9900773cd3/addon/-private/system/model/model.js#L739:L783
    - `DS.Model.reload`: https://github.com/emberjs/data/blob/b2a157b6500ea1cbf256ae2295096f9900773cd3/addon/-private/system/model/model.js#L739:L783
    - `DS.Model.eachRelationship`: https://github.com/emberjs/data/blob/b2a157b6500ea1cbf256ae2295096f9900773cd3/addon/-private/system/model/model.js#L1057:L1111
    - `DS.Model.eachRelatedType`: https://github.com/emberjs/data/blob/b2a157b6500ea1cbf256ae2295096f9900773cd3/addon/-private/system/model/model.js#L1715:L1733
    - `DS.Model.eachAttribute`: https://github.com/emberjs/data/blob/b2a157b6500ea1cbf256ae2295096f9900773cd3/addon/-private/system/model/model.js#L1863:L1910
    - `DS.Model.eachTransformedAttribute`: https://github.com/emberjs/data/blob/b2a157b6500ea1cbf256ae2295096f9900773cd3/addon/-private/system/model/model.js#L1912:L1960
    - `DS.RecordArray.update`: https://github.com/emberjs/data/blob/b2a157b6500ea1cbf256ae2295096f9900773cd3/addon/-private/system/record-arrays/record-array.js#L118:L155
    - `DS.HasManyReference.ids`: https://api.emberjs.com/ember-data/3.8/classes/DS.HasManyReference/methods/ids?anchor=ids
    - `DS.SnapshotRecordArray.snapshots`: https://api.emberjs.com/ember-data/3.8/classes/DS.SnapshotRecordArray/methods/snapshots?anchor=snapshots
    - `DS.Snapshot.belongsTo`: https://api.emberjs.com/ember-data/3.8/classes/DS.Snapshot/methods/belongsTo?anchor=belongsTo
    - `DS.Snapshot.hasMany`: https://api.emberjs.com/ember-data/3.8/classes/DS.Snapshot/methods/hasMany?anchor=hasMany
    - `DS.Snapshot.eachAttribute`: https://api.emberjs.com/ember-data/3.8/classes/DS.Snapshot/methods/eachAttribute?anchor=eachAttribute
    - `DS.Snapshot.eachRelationship`: https://api.emberjs.com/ember-data/3.8/classes/DS.Snapshot/methods/eachRelationship?anchor=eachRelationship
    - `DS.Snapshot.serialize`: https://api.emberjs.com/ember-data/3.8/classes/DS.Snapshot/methods/serialize?anchor=serialize
    - `DS.Store.query`: https://api.emberjs.com/ember-data/3.8/classes/DS.Store/methods/query?anchor=query
    - `DS.Store.queryRecord`: https://api.emberjs.com/ember-data/3.8/classes/DS.Store/methods/queryRecord?anchor=queryRecord
- [ ] ~Increase the version number in the header if appropriate.~
- [ ] ~If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.~

### CODE REVIEW

**Comment by 43081j** [comment]
You should probably follow convention here and name your type parameters with a `T` prefix: `TModel`, etc.

Also, if possible where your options are, a `Partial<{ option1: string; option2: number; }>` kinda solution would be nice instead of an entirely optional interface.

(build fails because `DS.Model` is redundant in that `Model` is in scope)

**Comment by chriskrycho** [comment]
@43081j thanks for commenting. `TModel` etc. is *not* a convention in TS in general or in this codebase in specific, and we're not going to adopt the C♯ conventions here. The fix here is indeed straightforward, though, and I'll have it pushed later or tomorrow.

**Comment by 43081j** [comment]
It was more to avoid confusion with the fact that there's multiple `Model` names going on (the type parameter and the class, used in the same statement). Many of the types in this repo do follow that convention, though you are right its not something DT necessarily pushes for (or checks for, afaik).

The partial options would be a nice one though at least so we can avoid over-matching unrelated interfaces.

**Comment by chriskrycho** [comment]
@43081j ah, to be clear, I meant specifically the Ember (including, here, ember-data) codebase!

As regards the partial options, those are as they are because that's literally the only thing specified by the docs; we're familiar with `Partial` but opt in/out in this particular codebase based on what is clearest for maintainers and consumers in any given spot.

**Comment by 43081j** [comment]
No worries I swooped in without fully understanding, they're fine being that they have only one option in most or all cases.

👍

File: types/ember-data/test/store.ts [hunk 1]
```diff
@@ -24,6 +24,7 @@ let post = store.createRecord('post', {
 });
 
 post.save(); // => POST to '/posts'
+post.save({ adapterOptions: { makeItSo: 'number one ' } });
```
Comment (mike-north) [inline types/ember-data/test/store.ts]: 👉 engage

**Review by mike-north — APPROVED** [review]

**Comment by rbuckton** [comment]
Thanks for the contribution!

**Pull request merged**